### ISSUE TITLE
Scheduling of a durable entity might be trigger up to 100ms before the scheduled time

### ISSUE THREAD

**Issue by havarnov** [Azure/azure-functions-durable-extension#1731]
### Description
When signaling a durable entity with a schedule time is says in the [documentation](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-entities?tabs=csharp#entity-operations):

> Scheduled time, which is an optional parameter for specifying the delivery time of the operation. For example, an operation can be reliably scheduled to run several days in the future.

### Expected behavior
The durable entity to be signaled exactly on the scheduled time or after.

### Actual behavior
The durable entity is signaled before the scheduled time.

### More info
I have looked at the source code and it seems like the durable entity can be called up to 100ms before the scheduled time:

https://github.com/Azure/azure-functions-durable-extension/blob/52169fcf925f332fc7537848af14ef9646cdc90e/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L779-L786

#### Example code

```cs
    public interface IScheduledEntity
    {
        Task Schedule();

        Task TriggeredOnScheduledTime(DateTime scheduledAt);
    }

    public class ScheduledEntity : IScheduledEntity
    {
        [FunctionName("ScheduledEntity.Run")]
        public static async Task Run(
            [EntityTrigger] IDurableEntityContext ctx)
        {
            await ctx.DispatchAsync<ScheduledEntity>();
        }

        public Task Schedule()
        {
            var scheduledTime = DateTime.UtcNow.AddMinutes(1);
            Entity.Current.SignalEntity<IScheduledEntity>(
                Entity.Current.EntityId,
                scheduledTime,
                n => n.TriggeredOnScheduledTime(scheduledTime));
            return Task.CompletedTask;
        }

        public Task TriggeredOnScheduledTime(DateTime scheduledAt)
        {
            if (DateTime.UtcNow < scheduledAt)
            {
                throw new Exception("Scheduled time should be after 'now'.");
            }

            return Task.CompletedTask;
        }
    }
```

### App Details

- **Azure Functions runtime version (1.0 or 2.0)**: 2.0
- **Programming language used**: csharp

**Comment by cgillum** [comment]
@sebastianburckhardt can you take a look at this and comment?

**Comment by sebastianburckhardt** [comment]
Looking at the code, indeed it is designed in a way such that there is a rare situation in which the message could be delivered early. For example, if setting the delivery time between (now + 7 days + 0ms) and (now + 7 days + 100ms). Possible but also quite unlikely. 

Another possibility, more likely, is that this could perhaps happen if the local clock differs from the Azure Storage clock; then the message would look early to the receiving node even though it is delivered at the correct time w.r.t. Azure Storage. I do not know if the Azure Queues make any particular guarantees about accuracy w.r.t the client clock.

@havarnov, did you observe this happening in a real situation?

If we think this is an issue we can add a little check in the code and delay for that little extra time.

**Comment by ConnorMcMahon** [comment]
This bug seems pretty minor to me, but the fix seems pretty straightforward. 

@sebastianburckhardt, could this just be fixed by changing `TimeSpan.FromMilliseconds(100)` to ` TimeSpan.FromMilliseconds(0)`. I think the timespan there was to account for clock drift, but I think I agree with @havarnov that we should lean towards executing a scheduled message after the scheduled time as opposed to before.

**Comment by havarnov** [comment]
@sebastianburckhardt I have experienced it in production code, but it's probably not due to the 7d + 100ms bug.

I have scheduled a durable entity with ```DateTime.UtcNow.AddMinutes(1.)```. So I guess it's due to clock drift between the Azure storage clock and "local" clock. If that's the case, then I guess that's a bug as well? IMO scheduling of messages must have some guarantee of precision. E.g. max clock drift of 100ms or something and should be documented as well.

**Comment by sebastianburckhardt** [comment]
@ConnorMcMahon, I think the solution you suggest is reasonable, especially because it is very simple.

My only concern with that solution is that if the local clock is behind Azure storage, this will likely stay that way and thus the message is *always* delivered early the first time, and thus going through the Queue twice *every time*. It seems potentially a bit wasteful. Still, it's probably o.k. and I think it is better to avoid adding more complexity to this path.

**Comment by havarnov** [comment]
@sebastianburckhardt and in addition to it being wasteful I guess it means that almost every messages will be delivered "later" than the current implementation.

**Comment by sebastianburckhardt** [comment]
Right. Perhaps it is better to not add any delay that is likely unnecessary for most applications. 

What we can do is document that "in the presence of clock discrepancies between the machines, scheduled signals may appear to arrive up to 100ms ahead of the scheduled time", and then suggest that if this is a problem, users insert a bit of wait time themselves:
```csharp
int millisecondsEarly = (int)(scheduledTime - DateTime.UtcNow).TotalMilliseconds;
if (millisecondsEarly > 0)
{
   await Task.Delay(millisecondsEarly);
}
```
